src/server/common/helpers/auth/fetch.js

```diff
@@ -0,0 +1,100 @@
+import Boom from '@hapi/boom'
+import fetch from 'node-fetch'
+import { getUserSession } from '#server/common/helpers/auth/get-user-session.js'
+
+export async function fetchWithAuthHeader(url, request) {
+  const accessToken = (await getUserSession(request)).token
+  const response = await fetch(url, {
+    headers: {
+      Authorization: `Bearer ${accessToken}`
+    }
+  })
+
+  const data = await response.json()
+
+  return { response, data }
+}
+
+function getOrganisationLinkHref(id, redirectUrl) {
+  return `/organisations/${id}/link?redirectUrl=${redirectUrl}`
+}
+
+export async function fetchWithAuthInterception(url, request, h) {
+  request.logger.info({ url }, 'auth: fetchWithAuthInterception')
+
+  try {
+    const { data, response } = await fetchWithAuthHeader(url, request)
+
+    if (!response.ok) {
+      throw Boom.unauthorized()
+    }
+
+    if (data.action === 'link-organisations') {
```

Comment:
This allows us to catch responses which have been "intercepted" by the API so that we can present an "Auth Wall" or "Takeover UI" whereby the user must confirm the link between the Defra ID Organisation in their token and an pEPR Organisation